R/asmodee.R

```diff
@@ -8,332 +8,411 @@
+
+  # loop through possible models until we can succesfully make prediction
+  i <- 0
+  success <- FALSE
+  tmp <- out$trending_model_fit
+  predict_catcher <- make_catcher(predict)
+  while(i < length(tmp) && !success) {
+    i <- i + 1
+    pred_result <- predict_catcher(
+      tmp[[i]],
+      new_data = data,
       alpha = alpha,
-      max_k = max_k,
-      method = method,
-      include_warnings = include_warnings,
-      ...
+      simulate_pi = simulate_pi,
+      uncertain = uncertain
     )
-    selected_model <- res_changepoint$model
-    selected_k <- res_changepoint$k
-  } else {
-    if (!is.numeric(fixed_k) |
-          !is.finite(fixed_k)) {
-      msg <- "`fixed_k` must be a finite number"
-      stop(msg)
+    names(pred_result) <- c("result", "prediction_warnings", "prediction_errors")
+    pred_result <- lapply(pred_result, list)
+    pred_result <- tibble::as_tibble(pred_result)
+
+    # remove prediction errors
+    keep <- vapply(pred_result$prediction_errors, is.null, logical(1))
+    pred_result <- pred_result[keep,]
+    if (nrow(pred_result) == 0) next
+
+    # remove prediction warnings (optionally)
+    if (!include_prediction_warnings) {
+      keep <- vapply(pred_result$prediction_warnings, is.null, logical(1))
+      pred_result <- pred_result[keep,]
     }
-    
-    selected_k <- as.integer(max(fixed_k, 0L))
-    data_train <- get_training_data(data, date_index, selected_k)
-
-    ## Here we need to eliminate models which would error when using predict on
-    ## the testing set; this can be due to new levels in the prediction set for
-    ## categorical predictors (factors), or to the presence of NAs in the
-    ## predictors.
-    models <- retain_sanitized_models(models,
-                                      data_train,
-                                      data,
-                                      warn = !quiet,
-                                      error_if_void = TRUE)
-    selected_model <- select_model(data_train, models, method, include_warnings, ...)
-    selected_model <- trending::fit(selected_model, data_train)
+
+    if (nrow(pred_result) == 0) next
+    success <- TRUE
   }
 
-  ## keep track of which data are training / testing and boundary dates
-  data <- set_training_data(data, date_index, selected_k)
-  training <- data$training
-  last_training_date <- max(dates[training], na.rm = TRUE)
-  first_testing_date <- NULL
-  if (selected_k > 0) {
-    first_testing_date <- min(dates[!training], na.rm = TRUE)
+  # error if no possible models
+  if (!success) {
+    if (include_prediction_warnings) {
+      msg <- "Unable to make prediction to the data without warnings or errors."
+    } else {
+      msg <- paste(
+        "Unable to make a prediction to the data without warnings or errors.",
+        "Consider using `include_prediction_warnings = TRUE`",
+        "to include more models which issued warnings.",
+        sep = "\n"
+      )
+    }
+    stop(msg)
```

Comment:
This may be a different situation from the previous one on fitting, so I think the switch is justified here.